.github/workflows/release.yml

```diff
@@ -1,30 +1,207 @@
 name: Release
 
 on:
+  # TODO: When a release PR is merged we should run a merge_group check to ensure:
+  # 1. The release PR is the only one in the merge group
+  # 2. The release PR is up-to-date with the target branch
+  #    i.e. no other PR has been merged since the release PR was added to the merge queue
+  workflow_dispatch:
   push:
     branches:
-      - main
+      - v-next
+
+defaults:
+  run:
+    shell: bash
 
 jobs:
   release:
     name: Release
     runs-on: ubuntu-latest
     permissions:
-      contents: write
-      pull-requests: write
+      contents: write # This allows us to push to the repository and create GitHub releases
+      pull-requests: write # This allows us to create pull requests
     steps:
       - name: Checkout Repo
         uses: actions/checkout@v4
         with:
           # This makes Actions fetch all Git history so that Changesets can generate changelogs with the correct commits
           fetch-depth: 0
 
-      - uses: ./.github/actions/setup-env
+      - name: Set up the environment
+        uses: ./.github/actions/setup-env
 
       - name: Install Dependencies
         run: pnpm install --frozen-lockfile --prefer-offline
 
       - name: Create Release Pull Request
+        id: pr
+        env:
+          # NOTE: If we use the GITHUB_TOKEN to create the release PR, the checks will not be triggered automatically
+          GITHUB_TOKEN: ${{ github.token }}
         uses: changesets/action@v1
+        with:
+          version: pnpm run version
+
+      - name: Build All Packages
+        if: steps.pr.outputs.hasChangesets == 'false'
+        run: pnpm run --recursive -no-bail --filter './v-next/**' --if-present build
+
+      - name: Publish All Packages (dry-run)
+        if: steps.pr.outputs.hasChangesets == 'false'
+        run: pnpm publish --filter "./v-next/**" -r --no-git-checks --tag next --access public --dry-run
+
+      # NOTE: When running publish on an already published package we expect it to succeed but not to release anything and not include + hardhat@ in the output.
+      # If it does, however, include + hardhat@ in the output, it's OK, because we'll just update the release without moving the release tag forward.
```

Comment:
It would be nice to know publish behaviour for already released packages. I haven't tested that scenario.